oioswift/common/middleware/container_hierarchy.py

```diff
@@ -650,28 +694,29 @@ def __call__(self, env, start_response):
             LOG.debug("%s: -> is NOT listing request", self.SWIFT_SOURCE)
             obj_parts = obj.split(self.DELIMITER)
             if len(obj_parts) > 1:
-                # path = self.DELIMITER.join(obj_parts[:-1]) + self.DELIMITER
                 path, _ = self._container_suffix(obj_parts, is_mpu)
-                is_dir = obj.endswith(self.DELIMITER)
-                # TODO (MBO) we should accept create key d1/d2/ only
-                # for empty object
+                if obj.endswith(self.DELIMITER):
+                    return self._handle_dir_object(req, start_response,
+                                                   account, container, path)
+
+                if not is_mpu and req.method in ('GET', 'HEAD'):
+                    # Hadoop S3 will made lot of HEAD requests on "directory"
+                    # like dir/subdirdir/ (which is managed by upper case)
+                    # but also dir/subdir: If the CNT KEY doesn't exist
+                    # for dir/ then subdir object does not exist
```

Comment:
I tried to fix the wording
> Hadoop S3 will **make** lot of HEAD requests on "directory" **objects**
> like dir/subdirdir/ (**with ending '/', handled by _handle_dir_object**)
> but also on dir/subdir **(without ending '/')**: if the CNT key doesn't exist
> for dir/, then subdir object won't exist **either**.

but I still don't understand.

Reply:
I've rewritten help, I hope it will be more clean

Reply:
Thanks.